File: ports/go/result/result.go
// Package result is the CiteNexus Result model and its parts
// (SPEC-PORTS-v1 §7). A Result is a grounded answer — or a refusal — carrying
// structured evidence signals and a reproducible provenance chain. It mirrors
// the Python reference citenexus.answer.result and serializes byte-compatibly
// with conformance/cases/result_roundtrip.json.
//
// Serialization contract: optional/absent scalar fields are nullable pointers so
// they marshal as null; list fields are initialized non-nil so they marshal as
// [] (never null or omitted).
package result

// RefusalAnswer is the single pinned refusal string every port must emit when it
// cannot ground an answer (§7).
const RefusalAnswer = "I can't answer that from the available evidence."

// Decision is the outcome recorded on the evidence signals.
type Decision string

const (
	DecisionAnswered Decision = "answered"
	DecisionRefused  Decision = "refused"
	DecisionPartial  Decision = "partial"
)

// TrustMode is the answer-flow trust mode. Only strict is exercised here.
type TrustMode string

const TrustModeStrict TrustMode = "strict"

// BBox is a page bounding box. Always null in the hermetic flow; present so the
// nullable field type is explicit.
type BBox struct {
	Page int     `json:"page"`
	X0   float64 `json:"x0"`
	Y0   float64 `json:"y0"`
	X1   float64 `json:"x1"`
	Y1   float64 `json:"y1"`
}

// EvidenceSignals are the structured retrieval/verification signals that replace
// a scalar confidence (§7/§12).
type EvidenceSignals struct {
	Decision                 Decision `json:"decision"`
	SupportingSources        int      `json:"supporting_sources"`
	DistinctDocuments        int      `json:"distinct_documents"`
	RetrievalScoreSpread     float64  `json:"retrieval_score_spread"`
	AllClaimsVerified        bool     `json:"all_claims_verified"`
	UnsupportedClaimsRemoved int      `json:"unsupported_claims_removed"`
	ConflictsDetected        int      `json:"conflicts_detected"`
	LanguagesInEvidence      []string `json:"languages_in_evidence"`
}

// SourceRef is a cited source: a verbatim passage in its source language, with an
// optional additive translation that never replaces the passage.
type SourceRef struct {
	Document        string  `json:"document"`
	Passage         string  `json:"passage"`
	PassageLanguage string  `json:"passage_language"`
	Page            *int    `json:"page"`
	BBox            *BBox   `json:"bbox"`
	SourceURI       *string `json:"source_uri"`
	Translation     *string `json:"translation"`
}

// Claim is a single claim in the answer and the evidence-unit ids that support it.
type Claim struct {
	Claim     string   `json:"claim"`
	Supported bool     `json:"supported"`
	Sources   []string `json:"sources"`
}

// ProvenanceEntry is one link of the reproducible chain: claim -> EU -> document
// -> S3 object -> checksum -> producing plugins.
type ProvenanceEntry struct {
	Claim        string         `json:"claim"`
	EvidenceUnit string         `json:"evidence_unit"`
	DocumentID   string         `json:"document_id"`
	S3Object     string         `json:"s3_object"`
	Checksum     string         `json:"checksum"`
	Page         *int           `json:"page"`
	BBox         *BBox          `json:"bbox"`
	ProducedBy   map[string]any `json:"produced_by"`
}

// Result is a grounded answer, or a refusal, with full evidence + provenance.
type Result struct {
	Answer          string            `json:"answer"`
	AnswerLanguage  string            `json:"answer_language"`
	Mode            TrustMode         `json:"mode"`
	Evidence        EvidenceSignals   `json:"evidence"`
	Claims          []Claim           `json:"claims"`
	Sources         []SourceRef       `json:"sources"`
	MissingEvidence []string          `json:"missing_evidence"`
	Conflicts       []string          `json:"conflicts"`
	Provenance      []ProvenanceEntry `json:"provenance"`
}

// Refused builds the pinned refusal Result in the given mode: refused decision,
// the pinned missing-evidence note, and empty (non-nil) claims/sources/provenance.
func Refused(mode TrustMode) Result {
	return Result{
		Answer:         RefusalAnswer,
		AnswerLanguage: "en",
		Mode:           mode,
		Evidence: EvidenceSignals{
			Decision:            DecisionRefused,
			LanguagesInEvidence: []string{},
		},
		Claims:          []Claim{},
		Sources:         []SourceRef{},
		MissingEvidence: []string{"no sufficiently relevant evidence found"},
		Conflicts:       []string{},
		Provenance:      []ProvenanceEntry{},
	}
}

File: src/citenexus/answer/result.py
"""The Result object and its parts (spec §16, §12, §11).

A Result is a grounded answer with a reproducible provenance chain. Confidence is
expressed as *structured signals* (§12), never a scalar — an uncalibrated "0.87"
is worse than none in regulated domains. The cited ``passage`` stays verbatim in
its source language; any ``translation`` is additive and never overwrites it (§11).
"""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict

from citenexus.domain.trust import TrustMode
from citenexus.evidence.unit import BBox


class Decision(StrEnum):
    """The outcome recorded on the evidence signals."""

    answered = "answered"
    refused = "refused"
    partial = "partial"


class EvidenceSignals(BaseModel):
    """Structured retrieval/verification signals — replaces scalar confidence (§12)."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    decision: Decision
    supporting_sources: int = 0
    distinct_documents: int = 0
    retrieval_score_spread: float = 0.0
    all_claims_verified: bool = False
    unsupported_claims_removed: int = 0
    conflicts_detected: int = 0
    languages_in_evidence: tuple[str, ...] = ()


class SourceRef(BaseModel):
    """A cited source: verbatim passage in its source language, plus optional
    additive translation (§16, §11)."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    document: str
    passage: str
    passage_language: str
    page: int | None = None
    bbox: BBox | None = None
    source_uri: str | None = None
    # Optional, MARKED translation in the answer language — never replaces passage.
    translation: str | None = None


class Claim(BaseModel):
    """A single claim in the answer and the EUs that support it."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    claim: str
    supported: bool
    sources: tuple[str, ...] = ()


class ProvenanceEntry(BaseModel):
    """One link of the reproducible chain: claim → EU → page+bbox → document →
    S3 object → checksum → producing plugins (§16)."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    claim: str
    evidence_unit: str
    document_id: str
    s3_object: str
    checksum: str
    page: int | None = None
    bbox: BBox | None = None
    # The artifact's ``produced_by`` stamp. Typed structurally for now; tightened
    # to the concrete ProducedBy model once `provenance-and-rebuild` lands (§4c).
    produced_by: dict[str, Any] | None = None


class Result(BaseModel):
    """A grounded answer, or a refusal, with full evidence + provenance."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    answer: str
    # The query language L. The answer is guaranteed to be in this language (§11);
    # independent of the evidence languages.
    answer_language: str
    mode: TrustMode
    evidence: EvidenceSignals
    claims: tuple[Claim, ...] = ()
    sources: tuple[SourceRef, ...] = ()
    missing_evidence: tuple[str, ...] = ()
    conflicts: tuple[str, ...] = ()
    provenance: tuple[ProvenanceEntry, ...] = ()
